### ISSUE TITLE
Pipeline node output artifacts should not be overridden if a node is executed multiple times

### ISSUE THREAD

**Issue by ptitzler** [elyra-ai/elyra#928]
Scenario:
 - Create a notebook pipeline.
 - Create a notebook (`abc.iynb`) that produces a unique result during each execution.
 - Add  the same [notebook] node twice to the pipeline.
 - Run the pipeline. 
 - Inspect the output artifacts `abc.ipynb` and `abc.html`

The outputs of notebook execution 1 are overridden by the outputs of notebook execution 2.

**Comment by lresende** [comment]
How did you run the pipeline?

Anyway, I believe this is not responsibility of the runtime, as we don't have control around that for local mode, and while running on remote runtime, then every run is isolated on a separate pod.

**Comment by kevin-bates** [comment]
I believe this is also an issue using the remote runtime. The overwrite doesn't occur when the node is processed, but rather when the outputs are uploaded to COS.  I _suspect_ this could be solved with more sophisticated input/output management where the node "ID" (or better yet, name) is somehow introduced into the equation - but that's pure conjecture.

This could also be resolved via documentation, and when things like looping constructs are supported, will just be what's expected - _"last writer wins"_.  Since I believe loop constructs are on the radar, we probably shouldn't spend time on this (other than perhaps documentation) unless it becomes a firm requirement.  What's important about pipelines is what comes out the end. :smile:

**Comment by ptitzler** [comment]
> The overwrite doesn't occur when the node is processed, but rather when the outputs are uploaded to COS

Yes, that's what I was trying to raise.

**Comment by lresende** [comment]
I probably know the answer here, but want to make sure we are on the same page. When you run two notebooks, with the same data, what are the things that can influence the different results?

**Comment by ptitzler** [comment]
> what are the things that can influence the different results?

Model training / hyperparameter tuning comes to mind as two tasks that one would run multiple times. (Multiple runs might produce different results because optimization might start with a random selection.) The other task is data loading. For example, our  tutorial pipelines use a universal script to load a data sets. 

> Looks like the ask here is really to scope not only the pipeline but each node execution with a timestamp,

Yep.



@nickp, wdyt?

**Comment by MLnick** [comment]
@ptitzler yes, hyperparameter tuning will generally have the same input data, with the only difference in runs being the input parameters to the ML pipeline / model. The results (trained model) would generally be different.

And yes also random init of models will lead to slightly different results each training run (unless a fixed random seed is used, which is not uncommon for reproducibility of pipelines).
